sceptre/hooks/__init__.py

```diff
@@ -61,16 +76,16 @@ def __get__(self, instance, type):
         """
         return getattr(instance, self.name)
 
-    def __set__(self, instance, value):
+    def __set__(self, instance: "Stack", value):
         """
         Attribute setter which adds a stack reference to any hooks in the
         data structure `value` and calls the setup method.
 
         """
 
-        def setup(attr, key, value):
-            value.stack = instance
-            value.setup()
+        def setup(attr, key, value: Hook):
+            attr[key] = clone = value.clone(instance)
+            clone.setup()
```

Comment:
This change actually patches a bug that we've had for a long time where the Stack instance on hooks inherited from parent StackGroups might not actually be the same as the current Stack. This ensures a fresh Hook is set up on each stack, even if its parameters have been inherited.